### ISSUE TITLE
Objects receive mouse events when canvas is obscured

### ISSUE THREAD

**Issue by sebastian-nowak** [paperjs/paper.js#992]
Quick demo:
```
var path = new Path.Circle({
    center: [100, 100],
    radius: 10,
    strokeColor: 'red',
    strokeWidth: 15
});
path.onMouseEnter = function(){ alert('huh?'); }

var div = document.createElement('div');
div.style.position = 'absolute';
div.style.top = 0;
div.style.bottom = 0;
div.style.left = 0;
div.style.right = 0;
div.style.background = 'rgba(0, 0, 0, .3)';
document.body.appendChild(div);
```
[The same code on http://sketch.paperjs.org](http://sketch.paperjs.org/#S/ZVHBagMhEP0V2Yu7ECSl9LKh9BByLBR66CGbg6uTKGucRceUEPLvHZv2ECKiznvPmTd6aaI+QtM3nxOQcc2iMWhrfNJJzJqceBURvsUHH9XaJxOgvQxR8DAQCVIvtk/L5ULwslvciKStL7ln6A/IlHCCNQZkuUxg5R3x5S05lr8M8dqthljLKozvWDJsag32sC/RkMfYdhehAyRqpSvuTXYrcR3iEKtd60+stGjKka0pk0ATbALUqJXMypqdd5XpHEDNmH3NyZekHjOGQiDvFIQzk8s7bEQiPD7AAfb0ACZ/cI/oqM10SFiirZXTYdQtP+Btqufu18J/EyPas9LzDNGunQ+25TTdir9p5O6mGX2k3PTb3fUH)

Tested on 0.9.25 and develop from Feb 3 (the one on http://sketch.paperjs.org). It happens even if the overlay isn't transparent at all.

**Comment by lehni** [comment]
Is this a bug? Wouldn't a DOM element behave the same way if the element that covers it doesn't intercept the event?

**Comment by sebastian-nowak** [comment]
Yes it is, only the DOM element on top receive the events. Try this code:
```
var path = new Path.Circle({
    center: [100, 100],
    radius: 10,
    strokeColor: 'red',
    strokeWidth: 15
});
path.onMouseEnter = function(){ alert('huh?'); }

var div = document.createElement('div');
div.style.position = 'absolute';
div.style.top = 0;
div.style.left = 0;
div.style.width = '300px';
div.style.height = '300px';
div.style.background = 'red';
div.style.zIndex = 1;
div.addEventListener('mouseenter', function(){ alert("won't fire"); });
document.body.appendChild(div);

var div = document.createElement('div');
div.style.position = 'absolute';
div.style.top = 0;
div.style.bottom = 0;
div.style.left = 0;
div.style.right = 0;
div.style.background = 'rgba(0, 0, 0, .3)';
div.style.zIndex = 2;
document.body.appendChild(div);
```
[The same code on http://sketch.paperjs.org](http://sketch.paperjs.org/#S/vZNNa4NAEIb/yuJlDQQxDb2klB5CDoUWCj30EHNY3UlcNDOyjvloyH/vbJIepE3bU0XUnfd1Ph7XQ4RmDdEkeq2AizIaRgXZsN4YrxrDpbpXCFv1Io/J1PmihviQoZKjAGTwEzUfpelQyWUxPAveWNe1EwldAi17qmBKNYlde7C6J7w5y6XYbzM8Du4yDGUTwmfqWpiFGtLDssOCHWE8OChTg+dYl135oAd36phhhqFd6zbitFR0a2ktKTwYhlkNYRVrUXXILvek5X0NSUOtCznlJW3yluqOQfccTI2IaS9Ww5K/BLdhhJBnnKbNrp+kBLcq+YqYm6JaeerQBkNA05PfH9HCTqTRJWysnW1knifXMiD4WK8DptOn0MNvMGXRllCzWjoPWRRwnSB8QsrJ7hPTNIB2WrraxlIkGP6TaE7MtP4baH9hmf5EcZWbWLbk+UzGgytMb34HIf9DLkNXDTnkNprMF8cP)

Here's a real life example of this issue:
![image](https://cloud.githubusercontent.com/assets/5895458/13261176/36c27c32-da5f-11e5-8226-f5f248028bc6.png)
Canvas below is receiving events while moving mouse above the modal.

**Comment by lehni** [comment]
Ok, I see what you mean. The problem is that the elements in the paper.js scene graph are not inserted in the DOM, only the canvas is, so we have to emulate mousenter / leave, through the mousemove events that the canvas receives, and those are still received. But we probably can install 'mouseenter' / 'mouseleave' events on the canvas and then only process these type of events if we're actually over the view or were just leaving it.

**Comment by sebastian-nowak** [comment]
I guess that would work for fixing `mouse*` events, but I think the real issue might be in the way `View._focused` is calculated. I don't know what was the intended behaviour, but currently it always points to the last view user had interaction with (unless it was removed).

Not only `mouse*` events are affected, but also the `key*`. There's something seriously wrong here, try typing in the input field in the top right corner.
```
function onKeyDown(event)
{
    alert('huh?');
}

var div = document.createElement('div');
div.style.position = 'absolute';
div.style.top = 0;
div.style.bottom = 0;
div.style.left = 0;
div.style.right = 0;
div.style.background = 'rgba(0, 0, 0, .3)';
document.body.appendChild(div);

var input = document.createElement('input');
input.setAttribute('type', 'text');
input.style.position = 'absolute';
input.style.top = 0;
input.style.right = 0;
div.appendChild(input);
```
[Link to http://sketch.paperjs.org](http://sketch.paperjs.org/#S/fZHBaoQwEIZfJXiJgshCb7sspbQ99dhj7SExs2tQZ0IcbWXZd2/MUqoVNgSS+edPZr7kkqDqINkn7w1wVSd5UpGZ49OAFVtCQfgG0wt9YQojIGclXkoUYagWPKeyHupHmR1KvJZY4qi8MHYUR2GoGrpwoKg8KIbXFuYolSEb7WEtep5aKBz1NpY6Cql0T+3AIFcOJheSu5WmiZm6jdzCiTeit+d6q2pVNWdPA5q5sj9rle5ycZvFQxZb+IXQZKZCOQdonmvbmjRcM1PciC26ge8wx3ykjruiB35i9lYH0lTy5EDmQjJ8r0x3H2fp+Xuepfofetl99GWH8N06dNo4ssh9sv/4vP4A)

And there's also a scenario where user clicks on an input, types something and then the input looses focus (e.g. via `someInput.blur()`). View will receive the `key*` events too, even though nothing should be focused.

**Comment by lehni** [comment]
Just like paper.js elements are not real DOM elements and thus are bound to behave differently, the canvas element does not receive or loose key focus (it's not an input element), and thus behaves differently there too. We actually receive the key events on the document object.

"fixing" all this may become a huge undertaking, and it may be going further than what can be expected to work of a canvas library. There are other ways to handle this, e.g. you could install events in your containers that hover over the canvas, and prevent the events from bubbling up if they are received there.

**Comment by lehni** [comment]
BTW, `View._focused` is not really the issue here, the issue is that keyboard focusing only works with elements that actually receive keyboard input.

**Comment by lehni** [comment]
Perhaps `tabIndex` can come to the rescue here... https://davidwalsh.name/tabindex-focus
